### PR TITLE
Webpack refactor

### PR DESCRIPTION
**Pull request by geraldarthur**
## Purpose

* Leaner, more configurable generator
* Optimized assets

## Pre-merge ToDos

- [x] QA w/ Visuals team

### CODE REVIEW

File: generators/app/templates/webpack.config.js [hunk 1]
```diff
@@ -1,32 +1,91 @@
...
   const isProd = p || env.production || process.env.NODE_ENV === "production";
   let wpconfig = {
     mode: isProd ? "production" : "development",
     entry: {
-      bundle: "./src/js/app.js",
+      bundle: path.join(__dirname, "src/js/app.js"),
     },
 
-    devtool: isProd ? "source-map" : false,
+    devtool: !isProd ? "source-map" : false,
```
Comment (ctbarna) [inline generators/app/templates/webpack.config.js]: On the site, we always generate a sourcemap: https://github.com/axioscode/platform-site/blob/master/webpack.browser.config.js#L36

File: generators/app/templates/webpack.config.js [hunk 2]
```diff
@@ -1,32 +1,91 @@
...
-//     },
+
+    plugins: [
+      new webpack.DefinePlugin({
+        "process.env.NAME": JSON.stringify(vizConfig.name || ""),
+      }),
+
+      new HtmlWebpackPlugin({
+        template: path.join(__dirname, "src/index.ejs"),
+        filename: path.join(__dirname, "dist/index.html"),
```
Comment (ctbarna) [inline generators/app/templates/webpack.config.js]: I don't think you need this.

File: generators/app/templates/src/index.ejs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  <head>
+    <meta charset="utf-8">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1, maximum-scale=1, user-scalable=no" />
+    <meta property="fullbleed" content="<%= htmlWebpackPlugin.options.isFullbleed %>" />
+    <meta property="slug" content="<%= htmlWebpackPlugin.options.slug %>" />
+    <meta property="apple-fallback" content="<%= htmlWebpackPlugin.options.appleFallback %>" />
+    <meta property="newsletter-fallback" content="<%= htmlWebpackPlugin.options.newsletterFallback %>" />
+    <link rel="icon" type="image/png" href="https://static.axios.com/img/axiosvisuals-favicon-128x128.png" sizes="128x128" />
+    <link rel="stylesheet" type="text/css" href="{{static 'css/main.css'}}">
```
Comment (ctbarna) [inline generators/app/templates/src/index.ejs]: This should be handled by MiniCssExtractPlugin

File: package.json [hunk 4]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "generator-axios",
+  "name": "generator-prometheus",
```
[the remark this answers is not in the document]
Comment (geraldarthur) [reply package.json]: yup yup! working title for now to allow for local development and comparison. yeoman generators need to be symlinked

**Comment by geraldarthur** [comment]
Closes #15

File: generators/app/templates/webpack.config.js [hunk 5]
```diff
@@ -1,32 +1,258 @@
+/* eslint-env node */
+
+const webpack = require("webpack");
 const path = require("path");
+const HtmlWebpackPlugin = require("html-webpack-plugin");
+const ScriptExtHtmlWebpackPlugin = require("script-ext-html-webpack-plugin");
+const MiniCssExtractPlugin = require("mini-css-extract-plugin");
+const UglifyJsPlugin = require("uglifyjs-webpack-plugin");
```
Comment (bjacobel) [inline generators/app/templates/webpack.config.js]: IIRC, Webpack 4 uglifies by default? (Edit: I see below now that you're passing custom options but using the new `optimize.minimize` config key - nice, nice)

Comment (geraldarthur) [reply generators/app/templates/webpack.config.js]: Yeahhh. I started without it but MiniCSSExtract needs it apparently. Here is their README section on [prod builds](https://github.com/webpack-contrib/mini-css-extract-plugin#minimizing-for-production) that lead me to adding [optimize css assets webpack plugin](https://github.com/NMFR/optimize-css-assets-webpack-plugin).

**Pull request merged**